src/dsl/dsl.jl

```diff
@@ -89,22 +91,33 @@ function desugar_tildes(expr)
     end
 end
 
-function resolve_gen_macros(expr, __module__)
+# Resolves Gen macros to globally referenced macro calls;
+# fully macroexpands all macros Gen does not recognize
+function resolve_or_expand_macros(expr, __module__)
     MacroTools.postwalk(expr) do e
-        # Resolve Gen macros to globally referenced macrocalls
-        if (MacroTools.@capture(e, @namespace_.m_(args__)) &&
-            m in DSL_MACROS && __module__.eval(namespace) == @__MODULE__)
-            macro_ref = GlobalRef(@__MODULE__, m)
-            line_num = e.args[2]
-            Expr(:macrocall, macro_ref, line_num, args...)
-        elseif (MacroTools.@capture(e, @m_(args__)) &&
-                m in DSL_MACROS && isdefined(__module__, m) &&
-                getfield(__module__, m) == getfield(@__MODULE__, m))
-            macro_ref = GlobalRef(@__MODULE__, m)
-            line_num = e.args[2]
-            Expr(:macrocall, macro_ref, line_num, args...)
-        else
-            e
+        if MacroTools.@capture(e, @namespace_.m_(args__) | @m_(args__)) # for any macro expression...
+            # if this is already a GlobalRef, the macro's name is stored in `m.name`
+            macroname = m isa GlobalRef ? m.name : m
+
+            if macroname in DSL_MACROS # if this is a Gen macro name (ie. @trace or @param)
+                # The macro is either from the local module `__module__` or explicitly specifies a `namespace`
+                if namespace === nothing
+                    mod = __module__
+                else
+                    mod =__module__.eval(namespace)
+                end
+
+                # if the macro is defined in the given module, and refers to the Gen macro
+                if isdefined(mod, macroname) && getfield(mod, macroname) == getfield(@__MODULE__, macroname)
+                    macro_ref = GlobalRef(@__MODULE__, macroname)
+                    line_num = e.args[2]
+                    return Expr(:macrocall, macro_ref, line_num, args...)
+                end
+            end
+            # if we get here, then this is not a Gen macro, so macroexpand it
+            return macroexpand(__module__, e)
```

Comment:
Can this macroexpansion yield additional `@trace` calls that you need to process as above? Or will it try to expand them too?

Reply:
@georgematheos I think I may not have been sufficiently clear in my suggestions yesterday about how best to implement this feature! I think we should avoid doing manual macro-expansion for the DML, because Julia can handle that on its own and thereby preserve all standard Julia semantics. It's not immediately clear to me that doing macro-expansion as part of a `postwalk` will actually preserve those semantics, and so I'd avoid that if possibld. I think it's only for the SML that we want to manually macro-expand the entire function body (after protecting all the Gen-specific macros by rewriting them). And so the macro-expansion code should go under `dsl/static.jl` rather than `dsl/dsl.jl`